crates/blockchain/fork_choice.rs

```diff
@@ -0,0 +1,259 @@
+use ethereum_rust_core::{
+    types::{Block, BlockHash, BlockHeader, BlockNumber},
+    H256,
+};
+use ethereum_rust_storage::{error::StoreError, Store};
+
+use crate::{
+    error::{self, InvalidForkChoice},
+    is_canonical,
+};
+
+/// Applies new fork choice data to the current blockchain. It performs validity checks:
+/// - The finalized, safe and head hashes must correspond to already saved blocks.
+/// - The saved blocks should be in the correct order (finalized <= safe <= head).
+/// - They must be connected.
+///
+/// After the validity checks, the canonical chain is updated so that all head's ancestors
+/// and itself are made canonical.
+///
+/// If the fork choice state is applied correctly, the head block header is returned.
+pub fn apply_fork_choice(
+    store: &Store,
+    head_hash: H256,
+    safe_hash: H256,
+    finalized_hash: H256,
+) -> Result<BlockHeader, InvalidForkChoice> {
+    if head_hash.is_zero() {
+        return Err(InvalidForkChoice::InvalidHeadHash);
+    }
+
+    // We get the block bodies even if we only use headers them so we check that they are
+    // stored too.
+
+    let finalized_res = if !finalized_hash.is_zero() {
+        store.get_block_by_hash(finalized_hash)?
+    } else {
+        None
+    };
+
+    let safe_res = if !safe_hash.is_zero() {
+        store.get_block_by_hash(safe_hash)?
+    } else {
+        None
+    };
+
+    let head_res = store.get_block_by_hash(head_hash)?;
+
+    if !safe_hash.is_zero() {
+        check_order(&safe_res, &head_res)?;
+    }
+
+    if !finalized_hash.is_zero() && !safe_hash.is_zero() {
+        check_order(&finalized_res, &safe_res)?;
+    }
+
+    let Some(head_block) = head_res else {
+        return Err(InvalidForkChoice::Syncing);
+    };
+
+    let head = head_block.header;
+
+    total_difficulty_check(&head_hash, &head, store)?;
```

Comment:
is there something to check here? forkchoice is a post-merge concept, and total difficulty is a pre-merge concept. After the merge, the total difficulty should stay the same.

Probably out of scope of this PR, I can create a ticket

Reply:
Already open: https://github.com/lambdaclass/lambda_ethereum_rust/issues/790

It does stay the same. This check was put there, I was told, to check that the new head was not pre-merge.